validator/src/main.rs

```diff
@@ -0,0 +1,261 @@
+use std::{
+    net::{IpAddr, SocketAddr},
+    process::exit,
+    sync::Arc,
+    time,
+};
+
+use anyhow::{anyhow, Context, Result};
```

Comment:
nit, take it or leave it: we use `bail!` and `ensure!` enough times here that i might import them as well? not a big deal though.